smartsim/_core/_install/redisaiBuilder.py

```diff
@@ -0,0 +1,287 @@
+
+        :return: True if onnx backend should be built
+        """
+        return "onnxruntime" in self.mlpackages
+
+    def build(self) -> None:
+        """Build RedisAI
+
+        :param git_url: url from which to retrieve RedisAI
+        :param branch: branch to checkout
+        :param device: cpu or gpu
+        """
+
+        # Following is needed to make sure that the clone/checkout is not
+        # impeded by git LFS limits imposed by RedisAI
+        os.environ["GIT_LFS_SKIP_SMUDGE"] = "1"
+
+        self.src_path.mkdir(parents=True)
+        self.build_path.mkdir(parents=True)
+        self.package_path.mkdir(parents=True)
+
+        retrieve(self.source, self.src_path, depth=1, branch=self.version)
+
+        self._prepare_packages()
+
+        for package in self.mlpackages.values():
+            self._patch_source_files(package.rai_patches)
+        cmake_command = self._rai_cmake_cmd()
+        build_command = self._rai_build_cmd
+
+        logger.info("Configuring CMake Build")
+        if self.verbose:
+            print(" ".join(cmake_command))
+        self.run_command(cmake_command, self.build_path)
+
+        logger.info("Building RedisAI")
+        if self.verbose:
+            print(" ".join(cmake_command))
+        self.run_command(build_command, self.build_path)
+
+        if self.platform.operating_system == OperatingSystem.LINUX:
+            self._set_execute(CONFIG.lib_path / "redisai.so")
+
+    @staticmethod
+    def _set_execute(target: pathlib.Path) -> None:
+        """Set execute permissions for file
+
+        :param target: The target file to add execute permission
+        """
+        permissions = os.stat(target).st_mode | stat.S_IXUSR
+        os.chmod(target, permissions)
+
+    @staticmethod
+    def _find_closest_object(
+        start_path: pathlib.Path, target_obj: str
+    ) -> t.Optional[pathlib.Path]:
+        queue = deque([start_path])
+        while queue:
+            current_dir = queue.popleft()
+            current_target = current_dir / target_obj
+            if current_target.exists():
+                return current_target.parent
+            for sub_dir in current_dir.iterdir():
+                if sub_dir.is_dir():
+                    queue.append(sub_dir)
+        return None
+
+    def _prepare_packages(self) -> None:
+        """Ensure that retrieved archives/packages are in the expected location
+
+        RedisAI requires that the root directory of the backend is at
+        DEP_PATH/example_backend. Due to difficulties in retrieval methods and
+        naming conventions from different sources, this cannot be standardized.
+        Instead we try to find the parent of the "include" directory and assume
+        this is the root.
+        """
+
+        for package in self.mlpackages.values():
+            logger.info(f"Retrieving package: {package.name} {package.version}")
+            target_dir = self.package_path / package.name
+            package.retrieve(target_dir)
```

Comment:
Incredibly minor performance thing if we care, but it looks like the iterations of this loop are totally independent. We could open up each of these io bound processes in a separate thread to make retrieving assets go a bit faster!

Reply:
So I went down too deep a rabbit hole for this, but it looks like we basically saturate the device's bandwidth for each package since the provider's bandwidth (whether GCP, Meta, Github, whatever) is crazy high. Given that, I think we keep this for simplicity of code over performance

Reply:
Fair enough! I'm not going to argue with simplicity!!